Refactor the Linker, Ranker, Recogniser and Pipeline

thobson88 opened the pull request:
Closes #276.

Major refactor. The main changes are:
 - new [subclass structure](https://github.com/Living-with-machines/T-Res/issues/276#issuecomment-2553515504) in the Recogniser, Ranker & Linker (and common `load` and `run` interface)
 - new [dataclasses](https://github.com/Living-with-machines/T-Res/issues/281#issuecomment-2481624048) representing data structures throughout the pipeline: `Mention`, `Sentence`, `Candidates`, `Predictions`, etc.
 - simplified [Pipeline](https://github.com/Living-with-machines/T-Res/issues/276#issuecomment-2553523890) class, with all business logic moved into the Linker

thobson88 committed:
Ignore .DS_Store files

thobson88 committed:
Refactor linking methods as Linker subclasses

All tests not marked as 'deezy' are passing

thobson88 committed:
Add RelDisamb Linker subclass & fix tests

thobson88 committed:
Move rel_params to the RelDisambLinker class

thobson88 committed:
Remove superfluous return value

thobson88 committed:
Refactor Ranker into subclasses

thobson88 committed:
Modify Ranker class hierarchy; fix unit tests

thobson88 committed:
Fix method_name in DeezyMatchRanker

thobson88 committed:
Fix integration tests

thobson88 committed:
Move train method to DeezyMatchRanker

thobson88 committed:
Fix docstrings

thobson88 committed:
Move deezy-specific code to Ranker subclass

thobson88 committed:
Update docstrings

thobson88 committed:
Refactor the Recogniser class

thobson88 committed:
Simplify variable names & rename module to avoid clash

thobson88 committed:
Simplify Pipeline constructor

thobson88 committed:
Remove superfluous return value

thobson88 committed:
Add Pipeline constructor unit test

thobson88 committed:
Remove inconsistency in treating place_wqid

thobson88 committed:
Remove superfluous return value

thobson88 committed:
Add dataclasses to represent ranking candidates

thobson88 committed:
Integrate dataclasses into the Ranker

thobson88 committed:
Integrate dataclasses into Linker & Pipeline

thobson88 committed:
Add method return types

thobson88 committed:
Move dataclasses to own module

thobson88 committed:
Add new dataclasses for Ranker & Linker output

thobson88 committed:
Update dataclasses and integrate into Linker & Pipeline

thobson88 committed:
Reorder dataclasses module

thobson88 committed:
Rename Ranker cache & load method; Edit docstrings

thobson88 committed:
Replace Ranker method_name() with class attribute

thobson88 committed:
Replace Linker method_name() with class attribute

thobson88 committed:
Rename Ranker string matching method

thobson88 committed:
Rename ranker run method argument

thobson88 committed:
Remove obsolete Linker method

thobson88 committed:
Improve Linker run method signature

thobson88 committed:
Add geo coords & Wikidata class fields in WikidataLink

thobson88 committed:
Add Recogniser dataclasses & run method

thobson88 committed:
Refactor pipeline to remove linking logic

thobson88 committed:
Move REL predict method call to Linker

thobson88 committed:
Handle empty candidates case

thobson88 committed:
Handle mentions with empty candidates

thobson88 committed:
Tidy up

thobson88 committed:
Add modular/stepwise pipeline methods

thobson88 committed:
Adapt REL model training to new dataclasses

thobson88 committed:
Adapt experiments to new dataclasses

thobson88 committed:
Add best_disambiguation_score method

thobson88 committed:
Add text() method in TextCandidates dataclass

thobson88 committed:
Rename Linker & Recogniser methods to load()

thobson88 committed:
Move dataclasses module to utils folder

thobson88 committed:
Add methods to Predictions dataclass

thobson88 committed:
Add static constructor in Ranker class

thobson88 committed:
Add static constructor in Linker class

thobson88 committed:
Fix bug in Linker __str__ method

thobson88 committed:
Use refactored Ranker & Linker in experiments

thobson88 committed:
Rename ner module

thobson88 committed:
Update pipeline run_disambiguation signature

thobson88 committed:
Fix import

thobson88 committed:
More robust guard clause in Linker training

thobson88 committed:
Distinguish empty predictions from empty candidates

thobson88 committed:
Fix tests in test_experiments.py

thobson88 committed:
Rename dataclasses

thobson88 committed:
Add pretty print method for Predictions

thobson88 committed:
Add pretty print methods for SentenceMentions

thobson88 committed:
Move dataclass tests to test_dataclasses.py

thobson88 committed:
Add pretty print methods for CandidateLinks

thobson88 committed:
Add pretty print method for Candidates

thobson88 committed:
Tweak pretty printing

thobson88 committed:
Update example notebook: basic pipeline

thobson88 committed:
Update example notebook: Deezy mostpopular

thobson88 committed:
Update example notebooks: Deezy REL

thobson88 committed:
Update example notebook: Pipeline modular

thobson88 committed:
Update example notebook: Perfect mostpopular

thobson88 committed:
Add guard clauses

thobson88 committed:
Update example notebook: Load & use NER model

thobson88 committed:
Remove obsolete comment

thobson88 committed:
Update app config

thobson88 committed:
Update app: run_ner endpoint

thobson88 committed:
Update app: run_candidate_selection

thobson88 committed:
Update app: run_disambiguation & pipeline

thobson88 committed:
Remove obsolete code

thobson88 committed:
Update ci.yml

thobson88 committed:
Update ci.yml

rwood-97 commented on t_res/geoparser/linking.py:
```diff
@@ -26,8 +25,397 @@ class Linker:
+        wkdt_class = self.linking_resources["entity2class"].get(wqid)
+        if wkdt_class:
+            return wkdt_class
+        return None
```
```suggestion
        wkdt_class = self.linking_resources["entity2class"].get(wqid, None)
        eturn wkdt_class
```

rwood-97 commented on t_res/geoparser/linking.py:
```diff
@@ -26,8 +25,397 @@ class Linker:
+        coords = self.linking_resources["wqid_to_coords"].get(wqid)
+        if coords:
+            return coords
+        return None
```
```suggestion
        coords = self.linking_resources["wqid_to_coords"].get(wqid, None)
        return coords
```

rwood-97 commented on t_res/geoparser/linking.py:
```diff
@@ -26,8 +25,397 @@ class Linker:
+        if matches.is_empty():
+            return self.empty_candidates(matches.mention, matches.ranking_method, place_of_pub_wqid, place_of_pub)
+
+        wikidata_links = []
+        candidate_links = []
+        for match in matches.matches:
+            if not isinstance(match, StringMatchLinks):
+                raise ValueError("Expected StringMatchLinks instance.")
+            
+            for wqid in match.wqid_links:
+                freq = self.linking_resources["mentions_to_wikidata"][match.variation][wqid]
+                wikidata_links.append(MostPopularLink(
+                    wqid,
+                    wkdt_class=self.wkdt_class(wqid),
+                    freq=freq)
+                )
+
+            candidate_links.append(CandidateLinks(
+                match.as_string_match(), 
+                wikidata_links, 
+            ))
```
move `wikidata_links=[]` inside `for match in matches.matches` loop.

thobson88 replied:
good catch!

thobson88 replied:
Fixed in 791280d1805e537356c98670892e6494cb8a9947

rwood-97 commented on t_res/geoparser/linking.py:
```diff
@@ -26,8 +25,397 @@ class Linker:
+        origin_coords = self.wkdt_coords(place_of_pub_wqid)
+
+        wikidata_links = []
+        candidate_links = []
```
+1 comment above, move `wikidata_links=[]` into for loop

rwood-97 commented on t_res/geoparser/linking.py:
```diff
@@ -26,8 +25,397 @@ class Linker:
+        ret = dict()
+        for link in wikidata_links:
+            
+            distance = min(max_on_gb, link.geodist if link.geodist is not None else max_on_gb)
```
just checking logic here is:
- if distance between link + place of pub is less than 1000km, store the distance (e.g. 80km)
- else just store 1000km
- also if distances is None, store 1000m

thobson88 replied:
Yes that's correct. Same logic as in the [earlier version](https://github.com/Living-with-machines/T-Res/blob/cff94256be88c0dabd15b98efb0daec315f695c7/t_res/geoparser/linking.py#L313) (but expressed differently) .

rwood-97 commented on t_res/geoparser/linking.py:
```diff
@@ -26,8 +25,397 @@ class Linker:
+            if distance == 0.0:
+                distance_score = 1.0
+            else:
+                distance = (max_on_gb if distance > max_on_gb else distance)
```
...
- if distance is more than 1000km, store 1000km
- else just store distance (e.g. 80km)

thobson88 replied:
(see previous)

rwood-97 commented on t_res/geoparser/linking.py:
```diff
@@ -26,8 +25,397 @@ class Linker:
+
+            final_score = 0.0
+            if link.geodist is not None:
+                final_score = round((relv_score + distance_score) / 2, 3)
```
```suggestion
            final_score = round((relv_score + distance_score) / 2, 3) if link.geodist else 0.0
```

thobson88 replied:
I tried this but it changes the logic when `link.geodist` is equal to `0.0` (as that evaluates to `False`), so I've gone instead with:
```
final_score = round((relv_score + distance_score) / 2, 3) if link.geodist is not None else 0.0
```

rwood-97 commented on t_res/geoparser/linking.py:
```diff
@@ -144,224 +530,150 @@ def __str__(self) -> str:
+                return self.empty_candidates(matches.mention, matches.ranking_method, place_of_pub_wqid, place_of_pub)
+        
+        wikidata_links = []
+        candidate_links = []
```
+1 comment above, move `wikidata_links=[]` within for loop

rwood-97 commented on t_res/geoparser/linking.py:
```diff
@@ -144,224 +530,150 @@ def __str__(self) -> str:
 
-        .. note::
+            # Mention-to-wikidata absolute relevance:
+            qcrlv_score = wikidata_link.freq
+            qcm2w_score = wikidata_link.normalized_score
```
what is qcrlv and qcm2w

thobson88 replied:
Good questions. These names are preserved from the `rank_candidates` method in [rel_utils.py](https://github.com/Living-with-machines/T-Res/blob/cff94256be88c0dabd15b98efb0daec315f695c7/t_res/utils/rel_utils.py#L207). I think `qc` is shorthand for Wikidata candidate (with a "Q" ID). Then `rlv` refers to relevance and `m2w` must be mentions-to-wikidata (i.e. the source of the normalized scores).

I've removed those variable names altogether and replaced with:
```python
            # Normalize absolute mention-to-Wikidata relevance by entity:
            candidate_score_1 = wikidata_link.freq / max_cand_freq
            # Average of string similarity and mention-to-Wikidata normalized relevance:
            candidate_score_2 = (wikidata_link.normalized_score + string_similarity) / 2

            # Average of two candidate scores, normalized between 0 and 0.9:
            score = ((candidate_score_1 + candidate_score_2) / 2) * 0.9
```

rwood-97 commented on t_res/geoparser/ner.py:
```diff
@@ -16,16 +16,225 @@
+            pred_ent["score"] = float(pred_ent["score"])
+            pred_ent["entity"] = pred_ent["entity"]
+            pred_ent = ner_utils.fix_capitalization(pred_ent, sentence)
+            predictions = ner_utils.aggregate_entities(pred_ent, lEntities)
```
I think this is rewriting predictions each time

rwood-97 replied:
also I really don't like the way this function is changing lEntitites in place (@tim this is not your fault I just wanted to say I don't like it)

thobson88 replied:
Agreed. I've captured this in a [new issue](https://github.com/Living-with-machines/T-Res/issues/283) as it's not part of this refactoring task.

rwood-97 commented on t_res/geoparser/ranking.py:
```diff
@@ -259,126 +174,185 @@ def load_resources(self) -> dict:
+            list(set(list(mention_df["score"].unique()))), reverse=True
+        )[:1]
+
+        mention_df = mention_df[mention_df["score"].isin(top_scores)]
```
make separate issue about this step

thobson88 replied:
I've replaced this with:
```python
mention_df = mention_df.query('score == score.max()')
```

thobson88 committed:
Granular test parameters

thobson88 committed:
Register test markers in pyproject.toml

thobson88 committed:
Move gazetteer to ByDistanceLinker subclass

thobson88 committed:
Remove ranker argument

thobson88 committed:
Simplify logic

thobson88 committed:
Simplify logic

thobson88 committed:
Refactor Linker run method

thobson88 committed:
Remove superfluous field from MentionCandidates

thobson88 committed:
Simplify logic

thobson88 committed:
Simplify logic in reldisamb score calculation

thobson88 committed:
Simplify logic in PartialMatchRanker

thobson88 committed:
Align test assertions with canonical resources

thobson88 committed:
Reorder Pipeline run method args

thobson88 committed:
Use test resources in pipeline test

thobson88 committed:
Handle empty candidates in Predictions dataclass

thobson88 committed:
Update comments

thobson88 committed:
Add method to access interim (prior) predictions

thobson88 committed:
Rename APIQuery class

thobson88 committed:
Update API app_template.py to new pipeline

thobson88 committed:
Add mkdocs site and autotranslate from Sphinx

thobson88 committed:
Fix navigation & rst mis-translation

thobson88 committed:
Change docs colour scheme

thobson88 committed:
Add navigation grid in docs homepage

thobson88 committed:
Fix docs homepage text

thobson88 committed:
Add dataclasses docstrings & docs page

thobson88 committed:
Add docstring to dataclasses module

thobson88 committed:
Consistent method parameter names

thobson88 committed:
Update deezy_processing docstrings & docs page

thobson88 committed:
Update get_data docstrings & docs page

thobson88 committed:
Add docs pages for geoparser modules

thobson88 committed:
Update ner_utils docstrings & docs page

thobson88 committed:
Update preprocess_data docstrings & docs page

thobson88 committed:
Update process_data docstrings & docs page

thobson88 committed:
Update process_wikipedia docstrings & docs page

thobson88 committed:
Update rel_e2e & rel_utils docstrings & docs page

thobson88 committed:
Add pages & navigation for utils.REL module

thobson88 committed:
Update entity_disambiguation docstrings & docs page

thobson88 committed:
Update mulrel_ranker docstrings & docs page

thobson88 committed:
Update utils.REL utils docstrings & docs page

thobson88 committed:
Update vocabulary docstrings & docs page

thobson88 committed:
Simplify page titles

thobson88 committed:
Update NER module docstrings & docs page

thobson88 committed:
Update ranking module docstrings & docs page

thobson88 committed:
Update linking module docstrings & docs page

thobson88 committed:
Update pipeline module docstrings & docs page

thobson88 committed:
Update docs workflow from sphinx to mkdocs

thobson88 committed:
Don't copy the command prompt character

Prevents clipboard copying of the prompt character in code snippets tagged as 'console'.

thobson88 committed:
Update installation & resources docs pages

thobson88 commented:
TODO: 
 - [x] attach lat-lon data for all predictions (i.e. pipeline outputs), not only the ones linked by distance
 - [ ] update README

thobson88 committed:
Partially update complete-tour docs page

thobson88 committed:
Hide unfinished mkdocs pages

thobson88 committed:
Merge pull request #285 from Living-with-machines/mkdocs

Update docs site & migrate to Material for Mkdocs

thobson88 committed:
Merge branch 'mkdocs' into 276-refactor

thobson88 committed:
Fix integration test with microtoponym

thobson88 committed:
Fix Pipeline integration test with microtoponym

thobson88 committed:
Fix Pipeline integration tests with empty candidates

thobson88 committed:
Move coords field into WikidataLinks dataclass

thobson88 committed:
Fix bugs in Linker

thobson88 committed:
Add static constructor for Recogniser

thobson88 committed:
Fix bug in Ranker

thobson88 committed:
Handle empty candidates in Predictions dataclass

thobson88 committed:
Add BatchJob class, tests & sample files

thobson88 committed:
Add batch-job executable script

thobson88 committed:
Fix microtoponym handling

thobson88 committed:
Fix empty candidates handling

thobson88 committed:
Add batch job API and executable script

thobson88 committed:
Add log level config parameter in batch job

thobson88 committed:
Add debug-level logging

thobson88 committed:
Avoid non-String input to NER run method

thobson88 committed:
Avoid non-String input to pipeline run_text_recognition

thobson88 committed:
Add BatchJob methods for place of pub info

thobson88 committed:
Fix place of publication handling in BatchJob

thobson88 committed:
Fix place of publication logging in BatchJob

thobson88 committed:
Handle case of empty list of WikidataLinks

thobson88 committed:
Fix bug in Candidates dataclass

thobson88 committed:
Fix bugs in REL linking & dataclasses

thobson88 committed:
Limit to compatible versions of transformers dependency

thobson88 committed:
Merge pull request #289 from Living-with-machines/287-transformers

Limit to compatible versions of transformers dependency

thobson88 committed:
Allow partial REL parameter config

thobson88 committed:
Add logic to favour place of pub if a candidate

thobson88 committed:
Handle empty predictions in BatchJob results

thobson88 committed:
Merge pull request #288 from Living-with-machines/286-batch-processing

Batch processing interface

thobson88 committed:
Merge branch '286-batch-processing' into 277-combined-score

thobson88 committed:
Add combined score linking logic

thobson88 committed:
Add error message

thobson88 committed:
Handle failure to compute great circle distance

thobson88 committed:
Set normalize=True in haversine call

thobson88 committed:
Fix missing return statement

thobson88 committed:
Add GPU device param to Recogniser init

thobson88 committed:
Add GPU device config to Linker & Recogniser

thobson88 committed:
Add linker model name in log message

thobson88 committed:
Set Deezy config to verbose logging

thobson88 committed:
Add Deezy model print statements

thobson88 committed:
Pass HuggingFace Dataset to recogniser pipeline

thobson88 committed:
Omit empty SentenceMentions in batch job

thobson88 committed:
Allow partial DeezyMatch parameter config

thobson88 committed:
Test non-default Deezy params in BatchJob

thobson88 committed:
Handle empty article text in BatchJob

thobson88 committed:
Mark test as resource-dependent

thobson88 committed:
Remove print statements

thobson88 committed:
Add log message for REL device

thobson88 committed:
Add log message for Recogniser device

thobson88 committed:
Add best coords methods in Predictions dataclass

thobson88 committed:
Update Ranker API to take a list of toponym mentions

thobson88 committed:
Add CombinedScores dataclass & integration test

thobson88 committed:
Add predicted coordinates in BatchJob CSV results

thobson88 committed:
Add combined scores test with default place of pub

thobson88 committed:
Merge pull request #291 from Living-with-machines/274-gpu

GPU support

thobson88 committed:
Write predictions.csv in batch job results

thobson88 committed:
Add input row index column to BatchJob predictions CSV

thobson88 committed:
Update docs: split Complete Tour into separate pages

thobson88 committed:
Add class diagrams in docs

thobson88 committed:
Add dataclasses page in docs

thobson88 committed:
Finish Batch Job docs & fix links

thobson88 committed:
Update ci.yml

Migrate from deprecated v2 to v4.
